lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -109,69 +115,68 @@ void addGuess(Inst *RHS, int MaxCost, std::vector<Inst *> &Guesses,
     TooExpensive++;
 }
 
-/* TODO call findCands instead */
-void findVars(Inst *Root, std::vector<Inst *> &Vars) {
-  // breadth-first search
-  std::set<Inst *> Visited;
-  std::queue<Inst *> Q;
-  Q.push(Root);
-  while (!Q.empty()) {
-    Inst *I = Q.front();
-    Q.pop();
-    if (!Visited.insert(I).second)
-      continue;
-    if (I->K == Inst::Var)
-      Vars.push_back(I);
-    for (auto Op : I->Ops)
-      Q.push(Op);
-  }
+// most naive prune algorithm ever
+bool prune (Inst *I, std::vector<Inst *> &ReservedInsts) {
+  if (instCount(I) >= MaxNumInstructions)
+    return false;
+  if (ReservedInsts.empty())
+    return false;
+  return true;
 }
 
 void getGuesses(std::vector<Inst *> &Guesses,
-                std::vector<Inst *> &Inputs,
+                const std::vector<Inst *> &Inputs,
                 int Width, int LHSCost,
-                InstContext &IC) {
+                InstContext &IC, Inst *PrevInst, Inst *PrevSlot,
+                int &TooExpensive) {
 
-  int TooExpensive = 0;
-  // start with the nops -- but not a constant since that is
-  // legitimately faster to synthesize using the special-purpose code
-  for (auto I : Inputs) {
-    for (auto V : matchWidth(I, Width, IC))
-      addGuess(V, LHSCost, Guesses, TooExpensive);
-  }
+  std::vector<Inst *> PartialGuesses;
+
+  std::vector<Inst *> Comps(Inputs.begin(), Inputs.end());
 
+  Comps.push_back(IC.getReservedInst(0));
   // TODO enforce permitted widths
   // TODO try both the source and dest width, if they're different
   std::vector<Inst::Kind> Unary = {
     Inst::CtPop, Inst::BSwap, Inst::Cttz, Inst::Ctlz
   };
   if (Width > 1) {
     for (auto K : Unary) {
-      for (auto I : Inputs) {
-        for (auto V : matchWidth(I, Width, IC)) {
-          auto N = IC.getInst(K, Width, { V });
-          addGuess(N, LHSCost, Guesses, TooExpensive);
+      for (auto I : Comps) {
+        if (I->K == Inst::ReservedInst && I->Width == 0) {
+          std::vector<int> UnaryArgWidths = {16,32};
```

Comment:
what's going on here? why 16 and 32?

Reply:
This needs discuss:
For unary operation on a instruction, like 
C = bswap %reservedinst.

The width of reservedinst is not ready yet, I assigned 16 and 32 temporally.

Reply:
this is fine for now, but:
- also support 64
- add a comment explaining this situation (and anything else similar in this patch)
- write a test case for this!!!

Reply:
Sure